### PR TITLE
Add Polarion report plugin

### PR DESCRIPTION
**Pull request by KwisatzHaderach**
You can now report test run results directly into Polarion, xunit file is left behind for additional uses (e.g. Report Portal).

### CODE REVIEW

**Comment by lukaszachy** [comment]
Minimal install failed because report-junit is missing - IMO polarion should not be part of minimal install so 'report-polarion' to be added as subpackage? I'm not sure if setup.py can do 'report-polarion' depends on 'report-junit' thing.

**Comment by lukaszachy** [comment]
Not related to this change but can I get nicer exception if I don't have a valid config file?
```python
  File "/home/lzachar/gits/revs-tmt/tmt/steps/report/__init__.py", line 54, in go
    phase.go()
  File "/home/lzachar/gits/revs-tmt/tmt/steps/report/polarion.py", line 42, in go
    import_polarion()
  File "/home/lzachar/gits/revs-tmt/tmt/export.py", line 61, in import_polarion
    from pylero.work_item import TestCase as PolarionTestCase
  File "/home/lzachar/.virtualenvs/revtmt/lib/python3.10/site-packages/pylero/work_item.py", line 1534, in <module>
    cfg = Configuration()
  File "/home/lzachar/.virtualenvs/revtmt/lib/python3.10/site-packages/pylero/base_polarion.py", line 56, in __init__
    raise PyleroLibException("The config files/ENV vars do not "
pylero.exceptions.PyleroLibException: The config files/ENV vars do not exist or are not of the correct format. Valid files are: /home/lzachar/.virtualenvs/revtmt/lib/python3.10/site-packages/pylero/etc/pylero.cfg, /home/lzachar/.pylero or .pylero

```

**Comment by KwisatzHaderach** [comment]
Well, you can file an issue on pylero, I'm not going to parse every single exception pylero can throw.
Also IMO this is descriptive enough, if you followed the guides for setting up pylero.

**Comment by lukaszachy** [comment]
> Well, you can file an issue on pylero, I'm not going to parse every single exception pylero can throw. Also IMO this is descriptive enough, if you followed the guides for setting up pylero.

You can capture ALL pylero exception (log.debug them) and raise tmt exception which will terminate tmt command correctly (not via traceback)

**Comment by KwisatzHaderach** [comment]
Now I see what you mean, lets fix it here, at least for the login so it's not breaking the tmt return codes.

**Review by lukaszachy — COMMENTED** [review]
Looks good however I'm not sure about the way how 'id' gets into Result.
How about adding it as an argument instead of reading it from the 'data' dict? 

API of Result is changed either way- currently it becomes required part of the 'data' dictionary (which needs to be documented as well)

**Comment by KwisatzHaderach** [comment]
From what I understood we wanted to make the ID the main test identifier across all systems and as such it makes much more sense to me to have it together with the data, although it doesn't matter really how the argument is passed, the code will look mostly the same.

**Comment by lukaszachy** [comment]
> From what I understood we wanted to make the ID the main test identifier across all systems and as such it makes much more sense to me to have it together with the data, although it doesn't matter really how the argument is passed, the code will look mostly the same.

Aren't 'data' more for the outcome of test execution? (duration, exit-status, note are there). It'd say 'id' is static and similar to test.name - it is not expected to change.

Personally I'd prefer to have `Result(data, test)`  and let result pull necessary info from the Test instance.

**Comment by KwisatzHaderach** [comment]
You are right, after I posted I checked again only to see name as an argument, I completely agree with passing whole test reference, with that we can pull anything else later without changing execute module.

File: tmt/steps/report/junit.py [hunk 1]
```diff
@@ -47,6 +47,39 @@ def duration_to_seconds(duration: Optional[str]) -> Optional[int]:
             f"Malformed duration '{duration}' ({error}).")
 
 
+def make_junit_xml(report: "tmt.steps.report.ReportPlugin") -> "junit_xml.TestSuite":
```
Comment (happz) [inline tmt/steps/report/junit.py:50]: Please, try the following patch:

```patch
diff --git a/tmt/steps/report/junit.py b/tmt/steps/report/junit.py
index 4e89efa..63e7fed 100644
--- a/tmt/steps/report/junit.py
+++ b/tmt/steps/report/junit.py
@@ -1,6 +1,6 @@
 import os
 import types
-from typing import Any, List, Optional, overload
+from typing import Any, List, Optional, cast, overload
 
 import click
 
@@ -12,6 +12,11 @@ DEFAULT_NAME = "junit.xml"
 
 junit_xml: Optional[types.ModuleType] = None
 
+# Thanks to import burried in a function, we don't get the regular missing annotations
+# but "name ... is not defined". Define a special type to follow the test suite instance
+# around, and if junit_xml ever gets annotations, we can replace it with provided type.
+JunitTestSuite = Any
+
 
 def import_junit_xml() -> None:
     """
@@ -47,7 +52,7 @@ def duration_to_seconds(duration: Optional[str]) -> Optional[int]:
             f"Malformed duration '{duration}' ({error}).")
 
 
-def make_junit_xml(report: "tmt.steps.report.ReportPlugin") -> "junit_xml.TestSuite":
+def make_junit_xml(report: "tmt.steps.report.ReportPlugin") -> JunitTestSuite:
     """ Create junit xml object """
     import_junit_xml()
     assert junit_xml
@@ -77,7 +82,7 @@ def make_junit_xml(report: "tmt.steps.report.ReportPlugin") -> "junit_xml.TestSu
         suite.test_cases.append(case)
 
     assert report.workdir is not None
-    return suite
+    return cast(JunitTestSuite, suite)
 
 
 class ReportJUnit(tmt.steps.report.ReportPlugin):
```

This should take resolve the "optionality" of the `junit_xml` import.

File: tmt/steps/report/junit.py [hunk 2]
```diff
@@ -76,34 +109,8 @@ def go(self, *args: Any, **kwargs: Any) -> None:
...
-                case.add_skipped_info(result.result)
-            elif result.result == "warn":
-                case.add_error_info(result.result)
-            # Passed state is the default
-            suite.test_cases.append(case)
-
-        assert self.workdir is not None
+        suite = make_junit_xml(self)
+
         f_path = self.opt("file", os.path.join(self.workdir, DEFAULT_NAME))
```
Comment (happz) [inline tmt/steps/report/junit.py:114]: mypy complains: `tmt/steps/report/junit.py:119: error: Argument 1 to "join" has incompatible type "Optional[str]"; expected "Union[str, PathLike[str]]"  [arg-type]`

That's correct, `self.workdir` is `Optional[str]` and can be `None`. I assume workdir cannot be `None` anymore, at this point in workflow (@lukaszachy am I right?), adding `assert self.workdir is not None` would narrow the type.

Comment (happz) [reply tmt/steps/report/junit.py:114]: Two more reports:

```
tmt/steps/report/junit.py:123: error: Item "None" of "Optional[Module]" has no attribute "to_xml_report_file"  [union-attr]
tmt/steps/report/junit.py:126: error: Item "None" of "Optional[Module]" has no attribute "TestSuite"  [union-attr]
```

Similar situation, `junit_xml` is `Optional[types.Module]`, adding `assert junit_xml is not None` would narrow the type - we *know* `junit_xml` is not `None` anymore, but mypy cannot infer that on its own.

File: tmt/steps/report/polarion.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        for result in self.step.plan.execute.results():
+            if not result.id:
+                raise tmt.utils.ReportError(
+                    f"Test Case {result.name} is not exported to Polarion, "
+                    "please run 'tmt tests export --how polarion' on it")
+            work_item_id, test_project_id = get_polarion_ids(
+                PolarionWorkItem.query(
+                    result.id, fields=['work_item_id', 'project_id']))
+
+            if test_project_id not in project_span_ids.attrib['value']:
```
Comment (happz) [inline tmt/steps/report/polarion.py:83]: `tmt/steps/report/polarion.py:83: error: Item "None" of "Optional[Element]" has no attribute "attrib"  [union-attr]`

That's correct, `project_span_ids` as a return value of `xml_tree.find()` is indeed `Optional[Element]`, and nothing tests whether it's a valid value before trying to access its attributes. `if project_span_ids is None: raise error` or `assert project_span_ids is not None` would help mypy, with the former giving us better chance to handle the error situation with a more helpful message or a completely different set of actions.

File: tmt/steps/report/polarion.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                project_span_ids.attrib['value'] += f',{test_project_id}'
+
+            test_properties = {
+                'polarion-testcase-id': work_item_id,
+                'polarion-testcase-project-id': test_project_id}
+
+            test_case = testsuite.find(f"*[@name='{result.name}']")
```
Comment (happz) [inline tmt/steps/report/polarion.py:90]: `tmt/steps/report/polarion.py:90: error: Item "None" of "Optional[Element]" has no attribute "find"  [union-attr]`

Same situation: `testsuite` comes from `xml_tree.find('testsuite')`, and may be `None`. There is no test making sure it's not `None`, either a condition or `assert` is needed to rule out this branch.

File: tmt/steps/report/polarion.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            properties_elem = ET.SubElement(test_case, 'properties')
```
Comment (happz) [inline tmt/steps/report/polarion.py:91]: `tmt/steps/report/polarion.py:91: error: Argument 1 to "SubElement" has incompatible type "Union[Element, None, Any]"; expected "Element"  [arg-type]`

Same situation: `test_case` comes from `testsuite.find(f"*[@name='{result.name}']")`, may be `None`. Also, thanks to previous error, `testsuite` itself may be `None`, and `None.find()` is unknown, adding `Any` to the mix of possible types of `test_case`.

**Comment by KwisatzHaderach** [comment]
@happz thanks for the suggestions and making me understand the reason behind all the new asserts

I also added '# type: ignore' for the last 2 errors as junit_xml does not have typing yet (and likely never will) and I guess there is no other way do solve this?

**Review by lukaszachy — COMMENTED** [review]
How about adding to setup.py? To make life for pip users easier.

```python
'report-polarion': ['junit_xml', 'pylero'],
```

**Review by lukaszachy — APPROVED** [review]
LGTM, I think it is time to the rebase.

**Comment by thrix** [comment]
@KwisatzHaderach note that typing for `export.py` was merged via #1298, so you will need to rebase this

**Comment by KwisatzHaderach** [comment]
mypy issue is caused by tmt/export.py having type for project id set as Optional[int] where it should be str @thrix this should probably be fixed by another patch? (I don't want to change part of export typing here)

**Comment by thrix** [comment]
@KwisatzHaderach I would fix it here, for whatever reason that issues was not present when we merged that PR. It is not visible here, I would fix it here, you are anyway editing that file right?

**Comment by KwisatzHaderach** [comment]
Frankly I'm just editing out the underscore in the function name because I put it there before not expecting to use the function elsewhere and here I am a month later, using it elsewhere.
Also I'm not seeing the issue on my machine when running mypy and have very little idea about typing.
I actually need this in 1.16 and pushing like 3 new patches just to fix something not related to this does not seem like a good idea to me.

**Comment by KwisatzHaderach** [comment]
@thrix your bad typing was even not consistent and had str in one place that would show your error

**Pull request merged**